dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblem.java

```diff
@@ -0,0 +1,148 @@
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimaps;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * A linkage error describing an invalid reference from {@code sourceClass} to {@code symbol}.
+ *
+ * @see <a href="https://jlbp.dev/glossary.html#linkage-error">Java Dependency Glossary: Linkage
+ *     Error</a>
+ */
+public abstract class LinkageProblem {
+
+  private final Symbol symbol;
+  private final ClassFile sourceClass;
+  private final String symbolProblemMessage;
+
+  /**
+   * A linkage error describing an invalid reference.
+   *
+   * @param symbolProblemMessage human-friendly description of this linkage error. Used in
+   *     conjunction with {@code symbol}, this value explains why we consider the reference to
+   *     {@code symbol} as a linkage error.
+   * @param sourceClass the source of the invalid reference.
+   * @param symbol the target of the invalid reference
+   */
+  LinkageProblem(String symbolProblemMessage, ClassFile sourceClass, Symbol symbol) {
+    this.symbolProblemMessage = Preconditions.checkNotNull(symbolProblemMessage);
+    Preconditions.checkNotNull(symbol);
+
+    // After finding symbol problem, there is no need to have SuperClassSymbol over ClassSymbol.
+    this.symbol =
+        symbol instanceof SuperClassSymbol ? new ClassSymbol(symbol.getClassBinaryName()) : symbol;
+    this.sourceClass = Preconditions.checkNotNull(sourceClass);
+  }
+
+  /** Returns the target symbol that was not resolved. */
+  public Symbol getSymbol() {
+    return symbol;
+  }
+
+  /** Returns the source of the invalid reference which this linkage error represents. */
+  public ClassFile getSourceClass() {
+    return sourceClass;
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    if (this == other) {
+      return true;
+    }
+    if (other == null || getClass() != other.getClass()) {
+      return false;
+    }
+    LinkageProblem that = (LinkageProblem) other;
+    return symbol.equals(that.symbol) && Objects.equals(sourceClass, that.sourceClass);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(symbol, sourceClass);
+  }
+
+  @Override
+  public final String toString() {
+    return formatSymbolProblem() + " referenced by " + sourceClass;
+  }
+
+  /**
+   * Returns the description of the problem on the {@code symbol}. This description does not include
+   * the {@code sourceClass}. This value is useful when grouping {@link LinkageProblem}s by their
+   * {@code symbol}s.
+   */
+  public String formatSymbolProblem() {
+    return symbol.toString() + " " + symbolProblemMessage;
+  }
+
+  /** Returns mapping from symbol problem description to the names of the source classes. */
+  public static ImmutableMap<String, ImmutableList<String>> groupBySymbolProblem(
```

Comment:
Feels like another instance of using a data structure to stand in for a more specific type. Also feels very specific to one use case. Perhaps the Map can be pushed into the class so instead there's a method that returns the source classes given a linkage problem.

Also, I expect Set<String> or Collection<String> fits better here than List<String> unless order is significant for some reason.